mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java

```diff
@@ -19,21 +19,342 @@
+      FileSystem fs = Util.getFs(resultPath, conf);
+      RemoteIterator<LocatedFileStatus> fileStatuses = fs.listFiles(resultPath, false);
+      Set<String> filesToKeep = new HashSet<>();
+      Set<String> files = new HashSet<>();
+      while (fileStatuses.hasNext()) {
+        LocatedFileStatus status = fileStatuses.next();
+        String fileName = status.getPath().getName();
+        files.add(resultPath.toString() + "/" + fileName);
+        if (fileName.endsWith(COMMITTED_PREFIX)) {
+          LOG.debug("Reading committed file {}", fileName);
+          ClosedFileData cfd = readCommittedFile(table.io(), resultPath.toString() + "/" + fileName);
+          DataFiles.Builder builder = DataFiles.builder(PartitionSpec.unpartitioned())
+              .withPath(cfd.fileName)
+              .withFormat(cfd.fileFormat)
+              .withFileSizeInBytes(cfd.length)
+              .withMetrics(cfd.serializableMetrics.metrics());
+          append = append.appendFile(builder.build());
+          filesToKeep.add(cfd.fileName);
+        }
+      }
+      append.commit();
+      LOG.info("Iceberg write is committed for {} with files {}", table, filesToKeep);
+      files.removeAll(filesToKeep);
+      LOG.debug("Removing unused files: {}", files);
+      files.forEach(file -> table.io().deleteFile(file));
+      cleanupJob(jobContext);
+    }
+  }
+
+  private static void createCommittedFileFor(FileIO io, ClosedFileData closedFileData) throws IOException {
+    OutputFile commitFile = io.newOutputFile(closedFileData.fileName + COMMITTED_PREFIX);
+    ObjectOutputStream oos = new ObjectOutputStream(commitFile.createOrOverwrite());
+    oos.writeObject(closedFileData);
+    oos.close();
+    LOG.debug("Iceberg committed file is created {}", commitFile);
+  }
+
+  private static ClosedFileData readCommittedFile(FileIO io, String committedFileLocation) throws IOException {
+    try (ObjectInputStream ois = new ObjectInputStream(io.newInputFile(committedFileLocation).newStream())) {
+      return (ClosedFileData) ois.readObject();
+    } catch (ClassNotFoundException cnfe) {
+      throw new IOException("Can not parse committed file: " + committedFileLocation, cnfe);
+    }
+  }
+
+  private static final class ClosedFileData implements Serializable {
+    private String fileName;
+    private FileFormat fileFormat;
+    private Long length;
+    private SerializableMetrics serializableMetrics;
+
+    private ClosedFileData(String fileName, FileFormat fileFormat, Long length, Metrics metrics) {
+      this.fileName = fileName;
+      this.fileFormat = fileFormat;
+      this.length = length;
+      this.serializableMetrics = new SerializableMetrics(metrics);
+    }
+  }
+
+  /**
+   * We need this class, since Metrics in not Serializable (even though it implements the interface)
+   */
+  private static final class SerializableMetrics implements Serializable {
+    private Long rowCount;
+    private Map<Integer, Long> columnSizes;
```

Comment:
Don't think HashMap is serializable. TreeMap is, FWIW.

Reply:
Actually HashMap is serializable as I have found out with implementing the readObject / writeObject methods. This made me thinking and found a better solution for this problem. See: #1430